[core] Bring cpu device module closer to cuda's.

kumpera opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103172

By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc @ptrblck @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

kumpera committed:
[core] Bring cpu device module closer to cuda's.

By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

[ghstack-poisoned]

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

wz337 commented on torch/cpu/__init__.py:
```diff
@@ -1 +1,85 @@
+    N.B. This class only exists to facilitate device-agnostic code
+
+    """
+    cur_stream : Optional['torch.cuda.Stream']
```
nit: import torch?

kumpera replied:
can't, it causes an import loop :(

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

wanchaol commented on torch/cpu/__init__.py:
```diff
@@ -1,6 +1,90 @@
     return torch._C._cpu._is_cpu_support_vnni()
+
+
+def is_available() -> bool:
```
shall we also add `set_device`?

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

wz337 commented on docs/source/cpu.rst:
```diff
@@ -0,0 +1,23 @@
+torch.cuda
```
nit: This should be torch.cpu?

kumpera replied:
yup

wz337 commented:
Noob question: is there any way to add a test for this?

wz337 reviewed (approved):
LGTM

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

wanchaol reviewed (approved):
nit: `set_device` is still not there so it might not be a fully device agnositic yet? I think `torch.cpu.set_device` should be accepting a integer similar to cuda but does nothing inside.

albanD commented:
FYI @heidongxianhua that should allow us to harmonize all the backend code in the different integrations we're adding!

albanD commented on torch/cpu/__init__.py:
```diff
@@ -1,6 +1,109 @@
+r"""
+This package implements abstractions found in ``torch.cuda``
```
Note that this is missing a few more things from the list at https://github.com/pytorch/pytorch/blob/999abd56a7be3d73f0d2e9ab06356c7281a3307d/torch/utils/backend_registration.py#L31

But that can be done in a follow up for sure.

kumpera replied:
yes, I plan to follow this up with Protocol'ization of this design so we can have tests for this.

kumpera committed:
Update on "[core] Bring cpu device module closer to cuda's."


By implementing some of the functionality used by CUDA we make
implementing device agnostic code a lot easier.

With this set of changes it's now possible to get FSDP wrap a trivial
module. FWD/BWD still TBD.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

kumpera commented:
@pytorchmergebot  merge

kumpera commented:
@pytorchmergebot merge
